### PR TITLE
DOC: Improve docstrings of sparse matrices

### PR DESCRIPTION
**Pull request by nmayorov**
I always was bothered that some methods in sparse matrices had just empty docstrings. So initially I tried to provide a docstring for each method.

I also changed method docstring in `dok_matrix` which weren't rendered reasonably by Sphinx (these are methods inherited from dict). I think farther improvements can be made in this direction, i.e. write docstrings more to the point for each method inherited from dict.

Besides adding missing docstrings, I think we can also make them generally better and more consistent. I suggest to work on all that in this PR.

### CODE REVIEW

**Comment by rgommers** [comment]
This looks like a good idea to me.

File: scipy/sparse/dok.py [hunk 1]
```diff
@@ -111,6 +111,41 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
...
+        """
+        return super(dok_matrix, self).values()
+
+    def setdefault(self, key, default=None):
+        """See setdefault method of Python dict.
+
+        This is the method of the underlying dict object.
+        """
+        super(dok_matrix, self).setdefault(key, default=default)
+
```
Comment (perimosocordiae) [inline scipy/sparse/dok.py]: I'm not sure that I like adding all these inherited methods just for the docstrings.

File: scipy/sparse/csr.py [hunk 2]
```diff
@@ -169,6 +169,7 @@ def tocsr(self, copy=False):
     tocsr.__doc__ = spmatrix.tocsr.__doc__
 
     def tocsc(self, copy=False):
+        """Convert this matrix to Compressed Sparse Column format."""
```
Comment (perimosocordiae) [inline scipy/sparse/csr.py]: Shouldn't this docstring be the same as `spmatrix.tocsc.__doc__` ?

File: scipy/sparse/bsr.py [hunk 3]
```diff
@@ -322,9 +323,11 @@ def __setitem__(self,key,val):
     ######################
 
     def matvec(self, other):
+        """Multiply matrix by vector."""
         return self * other
 
     def matmat(self, other):
+        """Multiply matrix by matrix."""
```
Comment (perimosocordiae) [inline scipy/sparse/bsr.py]: Multiply _this_ sparse matrix by another matrix.

File: scipy/sparse/bsr.py [hunk 4]
```diff
@@ -309,6 +309,7 @@ def diagonal(self):
     ##########################
 
     def getdata(self,ind):
+        """Not implemented method."""
```
Comment (perimosocordiae) [inline scipy/sparse/bsr.py]: This should be more explanatory (why isn't it implemented?) or just left out.

File: scipy/sparse/base.py [hunk 5]
```diff
@@ -587,13 +603,21 @@ def transpose(self, axes=None, copy=False):
...
         return self.tocsr().conj()
 
     def conjugate(self):
         return self.conj()
 
+    conjugate.__doc__ = conj.__doc__
+
     # Renamed conjtranspose() -> getH() for compatibility with dense matrices
     def getH(self):
+        """Hermitian transpose of a matrix."""
```
Comment (perimosocordiae) [inline scipy/sparse/base.py]: Add a "see also" for the dense matrix method?

**Review by perimosocordiae — COMMENTED** [review]
Overall I like this PR. Is there an automated check to make sure all relevant methods have docstrings?

File: scipy/sparse/dok.py (hunk 1 above)
Comment (rgommers) [inline scipy/sparse/dok.py]: agree, probably better to remove these dummy methods from this PR

**Comment by rgommers** [comment]
> Is there an automated check to make sure all relevant methods have docstrings?

not sure that is needed, seems a bit overkill. we don't do that for other modules either

**Comment by nmayorov** [comment]
I addressed what you mentioned.

The docstrings in sparse generally don't follow the proper convention, but usually they are enough to understand how to use methods. I can either continue expanding them or we can leave this PR as a minimum improvement and someone can continue later.

**Comment by perimosocordiae** [comment]
LGTM, merging.

**Comment by perimosocordiae** [comment]
I agree that the BSR `getdata` method can just be removed. The rest of the `getdata` methods were removed back in v0.9 [changelog](https://github.com/scipy/scipy/blob/master/doc/release/0.9.0-notes.rst#scipy-sparse), but we must have missed this one.

**Comment by rgommers** [comment]
thanks @nmayorov @perimosocordiae

**Pull request merged**